tests/ui/stubs.js

```diff
@@ -0,0 +1,227 @@
+var fs = require("fs");
+var http = require("http");
+var url = require("url");
+
+var httputil = require("selenium-webdriver/http/util");
+var portprober = require("selenium-webdriver/net/portprober");
+var promise = require("selenium-webdriver").promise;
+
+var stubdata = require("./stubdata");
+
+function Stubs() {
+    this.reset();
+}
+
+var headers = {
+    "Access-Control-Allow-Origin": "*",
+    "Access-Control-Allow-Headers": "Authorization,Content-Type",
+    "Access-Control-Allow-Methods": "GET,POST,PUT"
+};
+
+Stubs.prototype.init = function(driver) {
+    return promise.all([this._serveGpr(), this._serveGithub(), this._serveSupplementary()])
+        .then((ports) => promise.map(ports, (port) => "http://127.0.0.1:" + port))
+        .then((bases) => promise
+            .all(promise.map(bases, (base) => httputil.waitForUrl(base + "/healthz")))
+            .then(() => promise.fulfilled(bases))
+        )
+        .then((bases) => {
+            driver.getHealthz = (path) => driver.get(bases[0] + "/healthz");
+            driver.getGprPage = (path) => driver.get(bases[0] + "#" + path);
+            driver.getHealthz();
+            driver.executeScript("window.localStorage['github_api_endpoint'] = '" + bases[1] + "';");
+            driver.executeScript("window.localStorage['supplementary_api_endpoint'] = '" + bases[2] + "';");
+            driver.executeScript("window.localStorage['github_delays'] = '" +
+                JSON.stringify({POST_MERGE: 10, POST_PUSH: 10}) + "';");
+            driver.getHealthz();
+            return promise.fulfilled();
+        });
+};
+
+Stubs.prototype.queueSuccessfulRewriteHistory = function() {
+    this.rewriteHistoryQueue.push(true);
+};
+
+Stubs.prototype.queueUnsuccessfulRewriteHistory = function() {
+    this.rewriteHistoryQueue.push(false);
+};
+
+Stubs.prototype.queueSuccessfulSquashMerge = function() {
+    this.squashMergeQueue.push(true);
+};
+
+Stubs.prototype.getGithubRequest = function(path) {
+    return this.githubRequests[path];
+};
+
+Stubs.prototype.stub = function(pr) {
+    var pullPage = clone(pr.pullPage)
+    this.githubPathsToServe["/repos/" + pr.user + "/" + pr.repo + "/pulls/" + pr.id] = pullPage;
+    this.githubPathsToServe["/repos/" + pr.user + "/" + pr.repo + "/pulls/" + pr.id + "/commits"] = clone(pr.commitsPage);
+    this.githubPathsToServe["/repos/" + pr.user + "/" + pr.repo + "/issues/" + pr.id + "/comments"] = clone(pr.commentsPage);
+    this.githubPathsToServe["/repos/" + pr.user + "/" + pr.repo + "/pulls/" + pr.id + "/merge"] = clone(pr.mergePage);
```

Comment:
May be cleaner to be using template strings here given we can? Would save on the `+` headache..